projects/laji/src/app/shared/service/filter.service.ts

```diff
@@ -13,26 +15,44 @@ export class FilterService {
 
   /**
    * Filters the given value
-   * @param value array of values
+   * @param arr array of values
    * @param filterBy filter by these values
    * @param matching if true returns matching if false returns those not matching
    */
-  filter(value: any, filterBy: FilterByType, matching = true): any {
-    if (!Array.isArray(value) || !filterBy) {
-      return value;
+  filter<
+    T extends FilterBaseType,
+    K extends string[],
+    S extends SearchRecord<K>,
+    Element extends T | S
+  >(
+    arr: Element[], filterBy: FilterByType<T, K>, matching = true
+  ): Element[] {
+    if (!Array.isArray(arr)) {
+      return [arr];
     }
-    let needle: any = filterBy;
-    let properties: any = null;
+    if (!filterBy && typeof filterBy !== 'boolean') {
```

Comment:
`filterBy === false` instead of  `typeof filterBy !== 'boolean'` would be more readable at least for me.

Reply:
These are not equivalent. The if condition says that filterBy must be falsy, but not a boolean. The suggested change says that filterBy must be falsy and must be false (i.e. it has to be boolean).

Reply:
The aforementioned `this.filter([{a: 1}, {a: null}], false);` input should not pass type check, but it does, which is a bug... I think maybe it would be the best to move on for now though instead of spending more time strengthening the typing here :D

Reply:
Right, yeah I meant `filterBy !== false`. But idk, that's not objectively better. Maybe this wasn't the most productive review, let's just get on with it... :D